core/src/main/scala/org/apache/spark/scheduler/ResultTask.scala

```diff
@@ -66,7 +67,10 @@ private[spark] class ResultTask[T, U](
     _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
 
     metrics = Some(context.taskMetrics)
-    func(context, rdd.iterator(partition, context))
+    val itr = rdd.iterator(partition, context)
+    val result = func(context, itr)
+    val computedFullPartition = itr.isEmpty || rdd.storageLevel != StorageLevel.NONE
```

Comment:
hmm, doesn't this need to know whether we've computed the full partition for *all* the rdds involved in this task, not just the last one?  Eg., what if an earlier rdd is cached, but the final one is not.  Or what if there is a coalesce, so the coalesced partition is not computed fully, but the partitions of the original rdds are computed fully?

:(  kinda worried this is a critical issue ...